spring-integration-security/src/main/java/org/springframework/integration/security/context/SecurityContextCleanupChannelInterceptor.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.integration.security.context;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.support.ChannelInterceptorAdapter;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+import org.springframework.security.core.context.SecurityContext;
+
+/**
+ * The {@link ExecutorChannelInterceptor} implementation responsible for
+ * the {@link SecurityContext} cleanup.
+ * <p>
+ * Designed to work in tandem with {@link SecurityContextPropagationChannelInterceptor}.
+ * However the {@link MessageChannel} can be marked for {@link SecurityContext} cleanup
+ * with this interceptor, meanwhile not for {@link SecurityContext} propagation
+ * with {@link SecurityContextPropagationChannelInterceptor}.
+ *
+ * @author Artem Bilan
+ * @since 4.2
+ *
+ * @see SecurityContextPropagationChannelInterceptor
+ * @see SecurityContextCleanupAdvice
+ */
+public class SecurityContextCleanupChannelInterceptor extends ChannelInterceptorAdapter
+		implements ExecutorChannelInterceptor {
+
+	@Override
+	public Message<?> beforeHandle(Message<?> message, MessageChannel channel, MessageHandler handler) {
+		return message;
+	}
+
+	@Override
+	public void afterSendCompletion(Message<?> message, MessageChannel channel, boolean sent, Exception ex) {
```

Comment:
Consider the case:
````xml
<channel id="input">
    <queue/>
    <interceptors> 
	<beans:bean
					class="org.springframework.integration.security.context.SecurityContextPropagationChannelInterceptor"/>
    </interceptors>
</channel>
...
<channel id="output">
    <queue/>
</channel>
````

So, with `preSend()` on the `input` we propagate `SecurityContext` to that Scheduler's Thread.
And on the `send()` of the `output` channel we leave that Thread. 

So, nothing helps us here to free that thread from the current ctx for the future.

You have noticed good case with your second comment: we really can mark `<poller>` with `SecurityContextCleanupAdvice`. And we can address that in the documentation and its JavaDocs. But looks like nothing helps us with the `ExecutorChannel` on it `send()` part...